compiler/resolution/wrappers.cpp

```diff
@@ -156,20 +221,507 @@ static void      formalIsDefaulted(FnSymbol*  fn,
 
+static bool      formalDefaultIsCall(ArgSymbol* formal);
+
 static void      defaultedFormalApplyDefaultForType(ArgSymbol* formal,
-                                                    FnSymbol*  wrapFn,
+                                                    BlockStmt* wrapFn,
                                                     VarSymbol* temp);
 
 static void      defaultedFormalApplyDefaultValue(FnSymbol*  fn,
                                                   ArgSymbol* formal,
-                                                  IntentTag  intent,
-                                                  FnSymbol*  wrapFn,
+                                                  bool       addAddrOf,
+                                                  BlockStmt* wrapFn,
                                                   VarSymbol* temp);
 
 static void      insertWrappedCall(FnSymbol* fn,
                                    FnSymbol* wrapper,
                                    CallExpr* call);
 
+static Symbol* createDefaultedActual(FnSymbol*  fn,
+                                     ArgSymbol* formal,
+                                     CallExpr*  call,
+                                     BlockStmt* body,
+                                     SymbolMap& copyMap);
+
+// info is used to handle out-of-order named arguments
+// if there aren't any out-of-order arguments (as with promotion)
+// it can be NULL.
+static void addDefaultsAndReorder(FnSymbol *fn,
+                                CallExpr* call,
+                                CallInfo* info,
+                                std::vector<ArgSymbol*>& actualFormals,
+                                bool resolveNewCode) {
+
+  int numFormals = fn->numFormals();
+  std::vector<Symbol*> newActuals(numFormals);
+  std::vector<int8_t> newActualDefaulted(numFormals);
+
+  // Gather the actuals into newActuals with NULLs where
+  // we need to fill in a default. This also happens
+  // to address the need to reorder the actuals.
+  int i = 0;
+  for_formals(formal, fn) {
+    bool actualProvidedForFormal = false;
+    Symbol* actualSym = NULL;
+    if (info) {
+      int j = 0;
+      for_vector(ArgSymbol, arg, actualFormals) {
+        if (arg == formal) {
+          actualSym = info->actuals.v[j];
+        }
+        j++;
+      }
+    } else {
+      int j = 0;
+      for_actuals(actual, call) {
+        if (actualFormals[j] == formal) {
+          actualProvidedForFormal = true;
+          if (SymExpr* se = toSymExpr(actual))
+            actualSym = se->symbol();
+        }
+        j++;
+      }
+    }
+
+    newActuals[i] = actualSym;
+    newActualDefaulted[i] = !actualProvidedForFormal;
+
+    i++;
+  }
+
+  // Remove the actuals from the call
+  // (we'll add them back again in a moment)
+  for_actuals(actual, call) {
+    actual->remove();
+  }
+
+  // Create a copyMap to handle cases like
+  //   proc f(a, b=a)
+  // in which event, we can't refer to formal 'a' in the default value.
```

Comment:
Can you clarify the above comment? I see in your tests that the default value CAN refer to 'a'.
Is the comment a typo? Did you mean something else?